### PR TITLE
Binary download support

### PR DESCRIPTION
**Pull request by wmathurin**
New flag returnBinary (default false) in network plugin sendRequest method.

The native side of the network plugin will send back the base64 encoded body and the content type.

### CODE REVIEW

File: libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java [hunk 1]
```diff
@@ -83,7 +86,7 @@
     @Override
     public boolean execute(String actionStr, JavaScriptPluginVersion jsVersion, JSONArray args,
                            CallbackContext callbackContext) throws JSONException {
-        Action action = null;
+        Action action;
```
Comment (wmathurin) [inline libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java]: Getting rid of a compiler warning.

File: libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java [hunk 2]
```diff
@@ -194,7 +202,7 @@ private RestRequest prepareRestRequest(JSONArray args) throws UnsupportedEncodin
             }
             final JSONObject headerParams = arg0.optJSONObject(HEADER_PARAMS_KEY);
             final Iterator<String> headerKeys = headerParams.keys();
-            final Map<String, String> additionalHeaders = new HashMap<String, String>();
+            final Map<String, String> additionalHeaders = new HashMap<>();
```
Comment (wmathurin) [inline libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java]: Getting rid of a compiler warning.

File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestResponse.java [hunk 3]
```diff
@@ -155,32 +154,12 @@ public void consumeQuietly() {
 		return responseAsBytes;
 	}
 
-    /**
-     * Checks if the response has a body.
-     *
-     * @return True - if response body is present, False - otherwise.
-     */
-	public boolean hasResponseBody() {
```
Comment (wmathurin) [inline libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/RestResponse.java]: That method was bogus - it's not checking for a body is checking for a response header content-type header set to json.

File: libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java [hunk 4]
```diff
@@ -137,16 +148,13 @@ public void onSuccess(RestRequest request, RestResponse response) {
...
                             callbackContext.success();
                         }
                     } catch (Exception e) {
                         SalesforceHybridLogger.e(TAG, "Error while parsing response", e);
-                        if (response.isSuccess()) {
-                            callbackContext.success();
-                        } else {
-                            onError(e);
-                        }
+                        onError(e);
```
Comment (wmathurin) [inline libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java]: We couldn't not manipulate the response at all even though it was a successful http.
Should we still return success or fail?

**Comment by brandonpage** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

### Tests: 

Name | Classname | Time|
--- | --- | ---|
testLoadAccountMRUObjectsFromServer | com.salesforce.androidsdk.smartsync.manager.MetadataManagerTest | 5.782|
testLoadGlobalMRUObjectsFromServer | com.salesforce.androidsdk.smartsync.manager.MetadataManagerTest | 5.636|


<p align="right" data-meta="generated_by_ci/circleci">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Pull request merged**